content/en/BestPractices/protect-your-apis-with-rate-limiting.md

```diff
@@ -0,0 +1,130 @@
+
+What is wrong with it?
+
+```
+  app.post('/verify-phone number',(req, res) => {
+	  const { phone } = req.body;	
+    	  verify.call(phone)
+	  //...
+  })
+```
+
+Fraudsters can easily abuse this API endpoint for toll fraud due to the lack of rate limiting.
+
+### Fix
+
+Let’s fix that. Several npm packages can do this. But for this example, we will use `express-rate-limit`:
+
+- Import it first. 
+- Then define a custom rate limit for the API endpoint (`callLimiter`). 
+- Set a window to 1 hr (60 * 60 * 1000 ms).
+- Set max limits for that window as appropriate (10 attempts).
+- Optionally, provide a friendly error message to the blocked user (`'Too many verification attempts, please try again after an hour'`)
+- Finally, apply the `callLimiter` middleware we defined to the endpoint.
+
+```
+import rateLimit from 'express-rate-limit'
+
+//...
+
+const callLimiter = rateLimit({
+	windowMs: 60 * 60 * 1000, // 1 hour
+	max: 10, // Limit each IP to 10 phone calls per `window` (here, per hour)
+	message:
+		'Too many verification attempts, please try again after an hour',
+	standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
+})
+ 
+//...
+app.post('/verify-phone-number', callLimiter, (req, res) => {
+	const { phone } = req.body;	
+  	verify.call(phone)
+	//...
+})
+```
+
+## Other Scenarios
+  
+### Denial Of Service
+  
+Attackers may overload the API with a large number of requests. Thus, the API becomes unavailable to your users if it exceeds the API’s capacity to process requests. This affects the availability of your APIs.
+  
+### Bruteforce And Credential Stuffing Attacks
+  
+Attackers often brute force user credentials with lists of common passwords. Similarly, attackers also abuse the lack of rate limiting to perform credential stuffing attacks. In the case of credential stuffing attacks, attackers attempt to log in to user accounts using credentials obtained from other breaches or stolen credentials.
+ 
+### Data Scraping
+  
+Companies have policies that restrict data scraping from their application. That doesn’t guarantee protection against bad actors. Bad actors try to circumvent your measures by pulling information from your APIs at a very high rate.
+  
+### Fix
+  
+Use API rate limiting to protect against these kinds of attacks. Refer to the implementation in the previous scenario.
+ 
+## Some Considerations
+  
+Before you implement rate limiting for your APIs, also consider these factors:
+  
+### Will It Impact User Experience?
+  
+Well, it depends. You might want to give the user more freedom when using certain API functionalities. On the other hand, you might want to be stricter for abuse-prone APIs. This requires planning. Plan and tailor your rate limits to your use case before applying them. Choose the limits based on how many requests you expect from the client applications under normal conditions.

```

Comment:
@sourcecodereviewer can you include an example of when it makes sense to "give the user more freedom"

I think we can then split this section into 2 or even 3 paragraphs

Reply:
sure @mjang-cobalt 👍